### PR TITLE
[SV] Introduce SV for op

### PR DESCRIPTION
**Pull request by uenoku**
This introduces SV for operation e.g:
```mlir
    sv.for %i = %c0_i2 to %c-1_i2 step %c1_i2 : i2 {
      %RANDOM = sv.macro.ref.se< "RANDOM"> : i32
      %index = sv.array_index_inout %_RANDOM[%i] : !hw.inout<uarray<3xi32>>, i2
      sv.bpassign %index, %RANDOM : i32
    }
```
which corresponds to:
```verilog
 for (logic [1:0] i = 2'h0; i < 2'h3; i = i + 2'h1) begin
    _RANDOM[i] = `RANDOM;
 end
```

For the simplicity, sv.for has single induction block argument (`%i` in the example above). This PR changes HWMemSimImp and LowerSeqToSV to replace verbatim for loop with sv.for operation.

### CODE REVIEW

File: include/circt/Dialect/SV/SVGenerate.td [hunk 1]
```diff
@@ -50,3 +50,26 @@ def GenerateCaseOp : SVOp<"generate.case",
...
+def ForProceduralOp : SVOp<"for.procedural", [SingleBlock, NoTerminator,
+                        ProceduralOp, ProceduralRegion]> {
+  let summary = "System verilog for loop";
+
+  let description = [{
+    This operation defines System Verilog for loop.
+  }];
+
+  let regions = (region SizedRegion<1>:$body);
+  let arguments = (ins I32Attr:$lowerBound,
```
Comment (darthscsi) [inline include/circt/Dialect/SV/SVGenerate.td:63]: Attributes or values?

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 2]
```diff
@@ -3540,6 +3546,26 @@ LogicalResult StmtEmitter::visitSV(GenerateCaseOp op) {
   return success();
 }
 
+LogicalResult StmtEmitter::visitSV(ForProceduralOp op) {
+  emitSVAttributes(op);
+  // TODO: location info?
+  startStatement();
+  // for (genvar i = 0; i < SIZE; i++) begin
```
Comment (darthscsi) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:3553]: To be clear, a generate for loop and a for loop are different.

File: include/circt/Dialect/SV/SVGenerate.td [hunk 3]
```diff
@@ -50,3 +50,26 @@ def GenerateCaseOp : SVOp<"generate.case",
 
   let hasVerifier = true;
 }
+
+def ForProceduralOp : SVOp<"for.procedural", [SingleBlock, NoTerminator,
```
[the remark this answers is not in the document]
Comment (uenoku) [reply include/circt/Dialect/SV/SVGenerate.td:54]: Right, I meant non-procedural version of for loop to be generate for loop.

File: include/circt/Dialect/SV/SVGenerate.td (hunk 1 above)
Comment (uenoku) [reply include/circt/Dialect/SV/SVGenerate.td]: Thanks, I think it makes sense to use normal values as lower and upper values so values are used in the latest version

**Review by mikeurbach — APPROVED** [review]
I think the implementation looks good, I guess my only concern is tool compatibility. Do we expect any concerns with `for`? Also, it seems like we use `logic` unconditionally for the induction variable, would that be a problem? I guess the issues we've seen before are with `automatic logic`, so maybe it's fine.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 4]
```diff
@@ -3541,6 +3547,30 @@ LogicalResult StmtEmitter::visitSV(GenerateCaseOp op) {
...
 }
 
+LogicalResult StmtEmitter::visitSV(ForOp op) {
+  emitSVAttributes(op);
+  llvm::SmallPtrSet<Operation *, 8> ops;
+  startStatement();
+  auto inductionVarName = op->getAttrOfType<StringAttr>("hw.verilogName");
+  ps << "for (" << PP::ibox2 << "logic ";
+  ps.invokeWithStringOS([&](auto &os) {
+    emitter.emitTypeDims(op.getInductionVar().getType(), op.getLoc(), os);
```
Comment (fabianschuiki) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:3557]: Do we ever expect to see `X` or `Z` values in the index? Otherwise you could change `logic` to `bit` such that we only operate on two-valued integers.

File: lib/Dialect/SV/SVOps.cpp [hunk 5]
```diff
@@ -976,6 +977,111 @@ void OrderedOutputOp::build(OpBuilder &builder, OperationState &result,
...
+//===----------------------------------------------------------------------===//
+
+void ForOp::build(OpBuilder &builder, OperationState &result,
+                  int64_t lowerBound, int64_t upperBound, int64_t step,
+                  IntegerType type, StringRef name,
+                  llvm::function_ref<void(BlockArgument)> body) {
+  auto lb = builder.create<hw::ConstantOp>(result.location, type, lowerBound);
+  auto ub = builder.create<hw::ConstantOp>(result.location, type, upperBound);
+  auto st = builder.create<hw::ConstantOp>(result.location, type, step);
+  build(builder, result, lb, ub, st, name, body);
```
Comment (fabianschuiki) [inline lib/Dialect/SV/SVOps.cpp:991]: Ah just realized that the bounds and step size are SSA values. In that case `logic` makes sense since we've been using four-value logic everywhere else 🤔

File: lib/Dialect/SV/SVOps.cpp [hunk 6]
```diff
@@ -976,6 +977,111 @@ void OrderedOutputOp::build(OpBuilder &builder, OperationState &result,
...
+  APInt lb, ub, step;
+  if (matchPattern(op.getLowerBound(), mlir::m_ConstantInt(&lb)) &&
+      matchPattern(op.getUpperBound(), mlir::m_ConstantInt(&ub)) &&
+      matchPattern(op.getStep(), mlir::m_ConstantInt(&step)) &&
+      lb + step == ub) {
+    // Unroll the loop if it's executed only once.
+    op.getInductionVar().replaceAllUsesWith(op.getLowerBound());
+    replaceOpWithRegion(rewriter, op, op.getBodyRegion());
+    rewriter.eraseOp(op);
+    return success();
```
Comment (fabianschuiki) [inline lib/Dialect/SV/SVOps.cpp:1080]: Awesome 😎

File: lib/Dialect/SV/Transforms/HWMemSimImpl.cpp [hunk 7]
```diff
@@ -576,36 +559,55 @@ void HWMemSimImpl::generateMemory(HWModuleOp op, FirMemory mem) {
...
+          // Construct the following nested for loops:
+          // ```
+          //   for (int i = 0; i < mem.depth; i++) begin
+          //     for (int j = 0; j < randomMeg.size; j += 32)
+          //       randomMem[j+31:j] = `RANDOM
+          //     Memory[i] = randomMem[mem.dataWidth - 1: 0];
+          // ```
+          b.create<sv::ForOp>(
+              0, mem.depth, 1, outerLoopIndVarType, "i",
+              [&](BlockArgument outerIndVar) {
```
Comment (fabianschuiki) [inline lib/Dialect/SV/Transforms/HWMemSimImpl.cpp:589]: This is so much cleaner than the verbatim stuff above! SV is really getting closer to being an AST dialect 🎉

File: lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp [hunk 8]
```diff
@@ -516,7 +544,8 @@ FirRegLower::RegLowerInfo FirRegLower::lower(FirRegOp reg) {
   // For array registers, we annotate ram_style attributes if
   // `addVivadoRAMAddressConflictSynthesisBugWorkaround` is enabled so that we
   // can workaround incorrect optimizations of vivado. See "RAM address conflict
-  // and Vivado synthesis bug" issue in the vivado forum for the more detail.
+  // and Vivado synthesis bug" issue in the vivado forum for the more
+  // detail.
```
Comment (fabianschuiki) [inline lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp:548]: Possibly unrelated whitespace change?

**Review by fabianschuiki — APPROVED** [review]
LGTM! Love how this gets rid of a whole bunch of `sv.verbatim` and how it makes the SV dialect more like an AST. Very nice 🎉

**Comment by fabianschuiki** [comment]
To @mikeurbach's point: maybe we should hoist the loop variable declaration into the module body if the lowering options disallow automatic variables? I wouldn't be surprised if tools supported these inline for loop variable declarations but didn't support automatic logic for whatever reason, so this might just work in all tools that we care about 😬

File: lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp (hunk 8 above)
Comment (uenoku) [reply lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp:548]: Yep, thanks!

File: lib/Dialect/SV/SVOps.cpp (hunk 5 above)
Comment (uenoku) [reply lib/Dialect/SV/SVOps.cpp:991]: Yes, currently sv.for is capable of creating non-synthesizable loops.
